### ISSUE TITLE
Suggestion: consider relaxing the From<io::Error> bound in TryFutureExt::timeout

### ISSUE THREAD

**Issue by rjsberry** [async-rs/futures-timer#21]
Thanks for the updates to Futures 03!

I've been wanting to use `TryFutureExt::timeout` in places where either a `TryFuture` doesn't implement `From<io::Error>`, or for timing out regular futures.

Currently to do this you need to transform the regular future with `never_error` and then create an arbitrary error mapping in order to have the future type-check for `TryFutureExt::timeout`:

```rust
long_future()
    .never_error()
    .map_err(|_| io::Error::new(/* required for type-checking only */))
    .timeout(...)
```

Have there been any discussions about implementing a similar error type that wraps errors with the original future, timer errors, and elapsed errors (similar to `tokio-timer`)? Something like this would allow us to remove the `From<io::Error>` bound.

I'm happy to submit a PR but thought I'd check with you guys first to see what your thoughts on the matter were.

**Comment by yoshuawuyts** [comment]
>  Have there been any discussions about implementing a similar error type that wraps errors with the original future, timer errors, and elapsed errors (similar to tokio-timer)? Something like this would allow us to remove the From<io::Error> bound.

We haven't talked about this, but this does sound like a nice feature to have, so I'd def welcome a PR to add this!

**Comment by danieleades** [comment]
yeah i need something like this too. I need to timeout out TryFutures and return the result. I need to be able to tell the difference between a timeout, and any other kind of error.

so if i'm starting with a `Future<Output = Result<I, E>>` I need to return something like a `Future<Output = Result<I, Waited<E>>>` where
```rust
enum Waited {
    TimedOut,
    Error(inner_error),
    executor_error,
}
```

**Comment by danieleades** [comment]
I was able to do this-
```rust
async fn timeout_fut<F, I, E>(fut: F, timeout: Option<Duration>) -> Result<I, Waited<E>>
    where F: Future<Output = Result<I,E>> {

    if let Some(duration) = timeout {
        fut.map_err(|e| Waited::Inner(e)).timeout(duration).await
    } else {
        fut.map_err(|e| Waited::Inner(e)).await
    }
}

#[derive(Debug)]
pub(crate) enum Waited<E> {
    TimedOut,
    Inner(E),
   RuntimeError,
}

impl<E> From<std::io::Error> for Waited<E> {
    fn from(err: std::io::Error) -> Waited<E> {
        Waited::RuntimeError
    }
}
```

This only works because i don't need a From<std::io::Error> for any other errors, and i don't much care about the type of runtime error (if i get one).
It's a bit hacky though

**Comment by danieleades** [comment]
see #26 
I've implemented my above solution, more or less. This removes the From\<io::Error\> bound, and makes it easier to distinguish between a timeout and an inner error.

**Comment by rjsberry** [comment]
Hi @danieleades. I've had a quick look at your PR. This covers some of what was discussed in the review of #22, leading to the draft of #24 which resolves this issue and also #14.

**Comment by danieleades** [comment]
Hi @rjsberry, thanks for the heads up. I took a look at your pull request, quite a different approach!

- I'm not sold on the idea of extending Future, rather than TryFuture. If you extend future you are changing the return type from T to Option\<T>, which to me seems a little more surprising than Result\<T, E> to Result\<T, X\<E>>. Looking at the historical docs, it looks like this crate *did* extend Future and Stream and one point, but decided to change to the current implementation. Though I have to admit your way is less restrictive!
- are you concerned about the ergonomics of colliding with https://rust-lang-nursery.github.io/futures-api-docs/0.3.0-alpha.18/futures/future/trait.FutureExt.html (and the equivalent StreamExt trait)?

**Comment by rjsberry** [comment]
> I'm not sold on the idea of extending Future, rather than TryFuture

This has already been discussed and the consensus is that having the timer implementation panic favours extending Future.

> changing the return type from T to  Option\<T\>, which to me seems a little more surprising than Resul\t<T, E\> to Result\<T, X\<E\>\>

Option is placeholder as in the PR notes. My preference is to introduce a new error type than have timeout errors return `io::ErrorKind::TimedOut`.

**Comment by danieleades** [comment]
hi @rjsberry, I had some initial success playing around with (unstable) [trait specialisation](https://github.com/rust-lang/rfcs/blob/master/text/1210-impl-specialization.md).

My approach was to implement FutureExt such that for SomeFuture: Future<Output=x>, SomeFuture.timeout() -> impl Future<Output=Result<x, TimedoutError<()>> and then specialising SomeOtherFuture: TryFuture<Ok=x, Err=y> such that SomeOtherFuture.timeout() -> impl TryFuture<Ok=x, Err=TimedoutError<y>>.

Basically i used specialisation to flatten the nested result returned by the TryError in order to remove a layer of pointless indirection. No 'flattening' is done in the general case.

It's not compiling, and the errors i'm getting make zero sense. So either this isn't quite possible the way i'm trying to do it, or it's not sufficiently stablised to work properly in this case. But it *looks* really promising...

**Comment by yoshuawuyts** [comment]
Closed by #36.
